payu/envmod.py

```diff
@@ -109,4 +125,79 @@ def lib_update(required_libs, lib_name):
             return '{0}/{1}'.format(mod_name, mod_version)
 
     # If there are no libraries, return an empty string
-    return ''
\ No newline at end of file
+    return ''
+
+
+def setup_user_modules(user_modules, user_modulepaths):
+    """Run module use + load commands for user-defined modules"""
+
+    if 'MODULESHOME' not in os.environ:
+        print(
+            'payu: warning: No Environment Modules found; ' +
+            'skipping running module use/load commands for any module ' +
+            'directories/modulefiles defined in config.yaml')
+        return
+
+    # Add user-defined directories to MODULEPATH
+    for modulepath in user_modulepaths:
+        if not os.path.isdir(modulepath):
+            raise ValueError(
+                f"Module directory is not found: {modulepath}" +
+                "\n Check paths listed under `modules: use:` in config.yaml")
+
+        module('use', modulepath)
+
+    for modulefile in user_modules:
+        # Check module exists and there is not multiple available
+        module_subcommand = f"avail --terse {modulefile}"
+        output = run_cmd(module_cmd(module_subcommand)).stderr
+
+        # Extract out the modulefiles available
+        modules = [line for line in output.strip().splitlines()
+                   if not (line.startswith('/') and line.endswith(':'))]
+
+        if len(modules) > 1:
+            # Modules are used for finding model executable paths - so check
+            # for unique module -TODO: Could be a warning rather than an error?
```

Comment:
Yeah there will be multiple results if there are multiple versions of a module. For non-model modules that is probably fine. For models it's probably not desirable, e.g. a user specifying `access-om2` rather than a specific version `access-om2/2024.03.0`.

Not sure what to do about that. Probably best to insist on users specifying a version at this stage, so leave as error but the error message should tell them what to do: namely use a specific version.

Reply:
Yeah, I've added a longer log message about using specific version, or change what modules directories are used. So will leave as an error for now